File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        res = []
        
        def dfs(i, cur, total):
            if total == target:
                res.append(cur[:])  #create copy
                return
            if total > target or i == len(candidates):
                return
            #include candidates[i]
            cur.append(candidates[i])
            dfs(i+1, cur, total+candidates[i])
            cur.pop()
            
            #skip candidates[i]
            while i+1 < len(candidates) and candidates[i] == candidates[i+1]:
                i += 1
            dfs(i+1, cur, total)
        dfs(0,[], 0)
        return res

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<std::vector<int>> res;
        vector<int> current;
        dfs(candidates, target, 0, current, res);
        return res;
    }
    
private:
    void dfs(const std::vector<int>& candidates, int target, int index, std::vector<int>& current, std::vector<std::vector<int>>& res) {
            if (target == 0) {
                res.push_back(current);
                return;
            }

            if (target < 0 || index >= candidates.size()) {
                return;
            }

            for (int i = index; i < candidates.size(); ++i) {
                // Skip duplicates
                if (i > index && candidates[i] == candidates[i - 1]) {
                    continue;
                }

                current.push_back(candidates[i]);
                dfs(candidates, target - candidates[i], i + 1, current, res);
                current.pop_back();
            }
        }
};